### DISCUSSION TITLE
Specify the database on session execution to reuse same model for different databases?

### DISCUSSION THREAD

**Discussion by Diego-Hernandez-Moodys** [sqlalchemy/sqlalchemy#8480] [Usage Questions]
Say the following model exists (taken from the docs) under DEV & PROD databases on the same host 
```python
class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    name = Column(String)
    fullname = Column(String)
    nickname = Column(String)
```

If I don't specify an database at the engine construction, I can query both tables if I specify the database names using pure SQL (ignore the SQL injection risk for now)
```python
engine = sqlalchemy.create_engine("mysql://cred:cred@host")

statement = f"""
    SELECT
        new.*
    FROM DEV.{User.__tablename__} new
    LEFT JOIN PROD.{User.__tablename__} prev
        ON new.name = prev.name
    WHERE
        prev.name IS NULL
"""
Session = sqlalchemy.orm.sessionmaker()
Session.configure(bind = engine)
session = Session()
results = session.execute(statement)
// do something with results
session.close
```
 
Is there a sqlalchemy ORM equivalent? I understand I can create different engines & sessions for each specific database, but then how would I use both sessions in the same query? Or maybe I can use the "USE {database}" at specific calls?

```python
engine_dev = sqlalchemy.create_engine("mysql://cred:cred@host/DEV")
engine_prod = sqlalchemy.create_engine("mysql://cred:cred@host/PROD")

Session = sqlalchemy.orm.sessionmaker()
Session.configure(bind = engine_dev)
Session.configure(bind = engine_prod)
session = Session()
```

**Comment by zzzeek** [comment]
hi -


any kind of prefix with a dot is considered to be a "schema" in SQLAlchemy.  See https://docs.sqlalchemy.org/en/14/core/metadata.html#specifying-the-schema-name 

also re: the MySQL URL there, you should be able to get away with that URL that has no database portion in it, but I think reflection features, if you were using them, will not work like that (known limitation on our end).   So you might need to have a database portion in the URL regardless.  you still can use the "schema" feature because that just renders a prefix in the SQL statements.

**Reply by Diego-Hernandez-Moodys** [reply]
I figured the reflection would be a limitation given the current setup -- that's fine. I can perform the raw SQL, but I thought it would be nice to utilize SqlAlchemy as much as possible.

The documentation you linked lists the schema parameter for Sqlalchemy CORE, is that something I can modify in ORM?

**Reply by zzzeek** [reply]
yes when using the ORM, the ORM models map to Table objects.    the Table has a name and an optional schema.  you'd set both  https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#explicit-schema-name-with-declarative-table
